service-base/src/test/java/org/eclipse/hono/service/credentials/AbstractCredentialsServiceTest.java

```diff
@@ -0,0 +1,259 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.hono.service.credentials;
+
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.unit.Async;
+import io.vertx.ext.unit.TestContext;
+import io.vertx.ext.unit.junit.VertxUnitRunner;
+import org.eclipse.hono.service.credentials.CompleteBaseCredentialsService;
+import org.eclipse.hono.service.credentials.CompleteCredentialsService;
+import org.eclipse.hono.util.CredentialsConstants;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import java.net.HttpURLConnection;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Tests verifying behavior of {@link CredentialsService}.
+ *
+ */
+@RunWith(VertxUnitRunner.class)
+abstract public class AbstractCredentialsServiceTest {
```

Comment:
`public abstract`?

Should probably be renamed to `AbstractCompleteCredentialsServiceTest` because it tests all the optional functionality ...